docs: Add docs for new performance APIs (#10017)
getsentry/sentry-javascript: docs/v8-new-performance-apis.md

ecosystem, and to allow to use common auto instrumentation packages to be able to cover more ground in the ever-changing
JavaScript landscape.

Since the way that OpenTelemetry works differs from how the SDK used to work, this required some changes in order to be
compatible.

Note that for Browser- or Edge-based SDKs, we are not (yet) using OpenTelemetry for auto instrumentation. However, in
order to keep the SDKs isomorphic - especially for SDKs for Meta-Frameworks like Next.js, Sveltekit or Remix - we made
the decision to align the performance APIs for all JavaScript-based SDKs.

## The "old" Way of Manual Performance Instrumentation

Previously, there where two key APIs for adding manual performance instrumentation to your applications:

- `startTransaction()`
- `span.startChild()`

This showed the underlying data model that Sentry was originally based on, which is that there is a root **Transaction**
which can have a nested tree of **Spans**.

## The new model: Goodbye Transactions, Hello Spans Everywhere!

In the new model, transactions are conceptually gone. Instead, you will _always_ operate on spans, no matter where in

> the tree you are. Note that in the background, spans _may_ still be bundled up to transaction for the UI. However, this

Reviewer: l: we should clarify that "the UI" is in fact the Sentry product UI
```suggestion
the tree you are. Note that in the background, spans _may_ still be grouped into a transaction for the Sentry UI. However, this
```

happens transparently, and from an SDK perspective, all you have to think about are spans.

## The Span schema

Previously, spans & transactions had a bunch of properties and methods to be used. Most of these have been removed in
favor of a slimmer, more straightforward API, which is also aligned with OpenTelemetry Spans. You can refer to the table
below to see which things used to exist, and how they can/should be mapped going forward:

| Old name              | Replace with                                         |
| --------------------- | ---------------------------------------------------- |
| `traceId`             | `spanContext().traceId`                              |
| `spanId`              | `spanContext().spanId`                               |

> | `parentSpanId`        | Same                                                 |

Reviewer: l: "Same" as in unchanged? Then I'd prefer that as I previously confused Same with applying the replacement value from the line above (maybe a me problem 😅)

Reviewer: ```suggestion
| `parentSpanId`        | Unchanged                                                 |
```

| `status`              | TODO: new signature                                  |
| `sampled`             | `spanContext().traceFlags`                           |

> | `startTimestamp`      | `startTime` - note that this has a different format! |

Author: This will be sad, because this is not nice in OTEL, but I guess we have to do it 😬 they use a different format for timestamps :(

Reviewer: l: Can we add either a link explaining the new format or a 1-liner explaining the difference?

Author: yeah, will do that!

| `tags`                | Use `attributes` instead, or set tags on the scope   |
| `data`                | Use `attributes` instead.                            |
| `transaction`         | Removed                                              |
| `instrumenter`        | Removed                                              |
| `finish()`            | `end()`                                              |
| `end()`               | Same                                                 |
| `setTag()`            | `setAttribute()`                                     |
| `setData()`           | `setAttribute()`                                     |

> | `setStatus()`         | TODO: new signature                                  |

Reviewer 2: we need a long term strategy for status, might be notable to write about in it's own section.

Author: yeah, I'll go through all the "??" sections here tomorrow a bit 😓

| `setHttpStatus()`     | ???                                                  |
| `setName()`           | TODO: `updateName()`                                 |
| `startChild()`        | Call `Sentry.startSpan()` independently              |

> | `isSuccess()`         | ???                                                  |

Author: `isSuccess()` is not used by us (and really only checks the status), so I think we can deprecate/remove this.

| `toTraceparent()`     | ???                                                  |

> | `toContext()`         | ???                                                  |

Author: Looking through the codebase, I can't find a single occurrence of this (`toContext()`) - do we need this anymore at all?

Reviewer 3: Maybe downstream SDKs depend on this but I can't imagine a situation where this would be *needed*. Let's trash it.

| `updateWithContext()` | ???                                                  |
| `getTraceContext()`   | ???                                                  |

In addition, a transaction has this API:

| Old name                    | Replace with                                          |
| --------------------------- | ----------------------------------------------------- |
| `name`                      | Same                                                  |

> | `trimEnd`                   | Removed                                               |

Author: What should we do about `trimEnd`? 🤔 should we keep this somehow, or rework how this API works? I think with only spans this maybe also makes less sense overall, but not sure...?

Reviewer 3: trimEnd is just about idleTransactions. @mydea is there a way in the new performance API to inspect the child spans for a particular span and set its endTimestamp to the highest timestamp of the child spans? If so, we can drop this.

Author: I think as we keep a list of the child spans, we should be able to keep the highest end timestamp and use this for `idleSpan.end(highestTimestamp)` 🤔 So I'd say this should be possible!

In contrast to the other two methods, this does not take a callback and this does not make the created span the active
span. You can use this method if you want to create loose spans that do not need to have any children:

```js
Sentry.startSpan({ name: 'outer' }, () => {
  const inner1 = Sentry.startInactiveSpan({ name: 'inner1' });
  const inner2 = Sentry.startInactiveSpan({ name: 'inner2' });

  // do something

  // manually end the spans
  inner1.end();
  inner2.end();
});
```

No span will ever be created as a child span of an inactive span.

## Other Notable Changes

In addition to generally changing the performance APIs, there are also some smaller changes that this brings with it.

### Changed `SamplingContext` for `tracesSampler()`

Currently, `tracesSampler()` can receive an arbitrary `SamplingContext` passed as argument.
While this is not defined anywhere in detail, the shape of this context will change in v8. Going forward, this will mostly receive the attributes of the span, as well as some other relevant data of the span. Some properties we used to (sometimes) pass there, like `req` for node-based SDKs or `location` for browser tracing, will not be passed anymore.

> ### No more `undefined` spans

Reviewer: not new but just wanted to say again: big DX improvement!
